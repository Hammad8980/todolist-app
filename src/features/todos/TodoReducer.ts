import type { Task } from "./TodoTaskTypes";

export type Action =
  | { type: 'ADD_TASK'; payload: string }
  | { type: 'DELETE_TASK'; payload: number }
  | { type: 'TOGGLE_TASK'; payload: number }
  | { type: 'SET_TASK'; payload: Task[] };

export function todoReducer(state: Task[], action: Action): Task[] {
  switch (action.type) {
    case 'ADD_TASK':
      return [
        ...state,
        { id: Date.now(), title: action.payload, isCompleted: false, priority: 'p1' },
      ];
    case 'DELETE_TASK':
      return state.filter(todo => todo.id !== action.payload);
    case 'TOGGLE_TASK':
      return state.map(todo =>
        todo.id === action.payload ? { ...todo, isCompleted: !todo.isCompleted } : todo
      );
    case 'SET_TASK':
      return action.payload;
    default:
      return state;
  }
}