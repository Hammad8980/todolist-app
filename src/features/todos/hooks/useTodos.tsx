import { useState } from 'react';
import { type Task } from '../TodoTaskTypes';

export function useTodos() {
  const [todos, setTodos] = useState<Task[]>([]);

  const onAddTask = (title: string) => {
    if (title.trim() !== '') {
      setTodos([...todos, { id: Date.now(), title, isCompleted: false, priority: 'p1' }]);
    }
  };

  const onDeleteTask = (id: number) => {
    setTodos(todos.filter(todo => todo.id !== id));
  };

  const onToggleTask = (id: number) => {
    setTodos(
      todos.map(todo => (todo.id === id ? { ...todo, isCompleted: !todo.isCompleted } : todo))
    );
  };

  return {
    todos,
    onAddTask,
    onDeleteTask,
    onToggleTask,
  };
}
