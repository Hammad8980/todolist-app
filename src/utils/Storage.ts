// Save data to local storage
export const saveToStorage = <T>(key: string, data: T): void => {
  try {
    localStorage.setItem(key, JSON.stringify(data));
  } catch (error) {
    console.error('Error saving to localStorage:', error);
  }
};

// Load data from local storage
export const loadFromStorage = <T>(key: string, defaultValue: T): T => {
  try {
    const storedData = localStorage.getItem(key);
    return storedData ? JSON.parse(storedData) : defaultValue;
  } catch (error) {
    console.error('Error loading from localStorage:', error);
    return defaultValue;
  }
}