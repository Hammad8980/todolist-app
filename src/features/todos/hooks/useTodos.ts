import { useReducer, useEffect } from 'react';
import { type Task } from '../TodoTaskTypes';
import { saveToStorage, loadFromStorage } from '../../../utils/Storage';

const STORAGE_KEY = 'Todo-App-Tasks';

type Action =
  | { type: 'ADD_TASK'; payload: string }
  | { type: 'DELETE_TASK'; payload: number }
  | { type: 'TOGGLE_TASK'; payload: number };

function todoReducer(state: Task[], action: Action): Task[] {
  switch (action.type) {
    case 'ADD_TASK':
      return [
        ...state,
        { id: Date.now(), title: action.payload, isCompleted: false, priority: 'p1' },
      ];
    case 'DELETE_TASK':
      return state.filter(todo => todo.id !== action.payload);
    case 'TOGGLE_TASK':
      return state.map(todo =>
        todo.id === action.payload ? { ...todo, isCompleted: !todo.isCompleted } : todo
      );
    default:
      return state;
  }
}

export function useTodos() {
  const [todos, dispatch] = useReducer(todoReducer, loadFromStorage<Task[]>(STORAGE_KEY, []));
  useEffect(() => {
    console.log('Saving todos...');
    try {
      saveToStorage(STORAGE_KEY, todos);
      console.log('Todos: ', todos);
    } catch (error) {
      console.log(error);
    }

    return () => {
      console.log('Cleaning up: effect is being cleaned');
    };
  }, [todos]);

  const onAddTask = (title: string) => {
    if (title.trim() !== '') {
      dispatch({ type: 'ADD_TASK', payload: title });
    }
  };

  const onDeleteTask = (id: number) => {
    dispatch({ type: 'DELETE_TASK', payload: id });
  };

  const onToggleTask = (id: number) => {
    dispatch({ type: 'TOGGLE_TASK', payload: id });
  };

  return {
    todos,
    onAddTask,
    onDeleteTask,
    onToggleTask,
  };
}
