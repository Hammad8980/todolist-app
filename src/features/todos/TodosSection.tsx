import TodoInput from '../../components/todo/TodoInput';
import TodoList from '../../components/todo/TodoList';
import { useTodos } from './hooks/useTodos';

export default function TodosSection() {
  const { todos, onAddTask, onDeleteTask, onToggleTask } = useTodos();

  return (
    <div>
      <TodoInput onAddTask={onAddTask} />

      <h2 className="text-lg font-bold mt-8 mb-4">Your list: </h2>

      {todos.map((task) => (
        <TodoList
          key={task.id}
          task={task}
          onDelete={onDeleteTask}
          onToggle={onToggleTask}
        />
      ))}
    </div>
  );
}
