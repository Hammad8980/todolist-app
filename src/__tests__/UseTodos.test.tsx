import { renderHook, act } from '@testing-library/react';
import { useTodos } from '../features/todos/hooks/useTodos';
import { todoReducer, type Action } from '../features/todos/TodoReducer';
import * as Storage from '../utils/Storage';
import type { Task } from '../features/todos/TodoTaskTypes';


// Mock Storage functions
const STORAGE_KEY = 'Todo-App-Tasks';

describe('useTodos', () => {
  beforeEach(() => {
    jest.useFakeTimers();
    jest.spyOn(Storage, 'loadFromStorage').mockImplementation(() => []);
    jest.spyOn(Storage, 'saveToStorage').mockImplementation(() => {});
  });

  afterEach(() => {
    jest.clearAllMocks();
    jest.clearAllTimers();
    jest.useRealTimers();
  });

  describe('TodoReducer', () => {
    const initialState: Task[] = [
      {
        id: 1,
        title: 'Initial Task',
        isCompleted: false,
        priority: 'p1',
      },
    ];
    it( 'should add a task', () => {
        const result = todoReducer(initialState, { type: 'ADD_TASK', payload: 'New Task' })
        expect(result.length).toBe(2);
    })
    it('should delete a task', () => {
        const result = todoReducer(initialState, { type: 'DELETE_TASK', payload: 1 })
        expect(result.length).toBe(0);
    })
    it('should toggle a task', () => {
        const result = todoReducer(initialState, { type: 'TOGGLE_TASK', payload: 1})
        expect(result[0].isCompleted).toBe(true);
    })
    it('should set tasks', () => {
        const newTasks:Task[] = [{
            id:2,
            title:'My new task',
            isCompleted:false,
            priority:'p2'
        }]
        const result = todoReducer(initialState, {type: 'SET_TASK', payload: newTasks})
        expect(result).toEqual(newTasks);

    })
    it('should return current state for unkown actions', () => {
        const result = todoReducer(initialState, { type: 'UNKNOWN'} as unknown as Action);
        expect(result).toBe(initialState);
    });
  });

  it('should initialize with loading state', () => {
    const { result } = renderHook(() => useTodos());
    expect(result.current.isLoading).toBe(true);
    expect(result.current.todos).toEqual([]);
    expect(result.current).toMatchSnapshot();
  });

  it('should load todos from storage after timeout', async () => {
    const mockTodos: Task[] = [{ id: 1, title: 'Test', isCompleted: false }];
    (Storage.loadFromStorage as jest.Mock).mockReturnValueOnce(mockTodos);

    const { result } = renderHook(() => useTodos());
    act(() => {
      jest.runAllTimers();
    });

    expect(Storage.loadFromStorage).toHaveBeenCalledWith(STORAGE_KEY, []);
    expect(result.current.todos).toEqual(mockTodos);
    expect(result.current.isLoading).toBe(false);
    expect(result.current).toMatchSnapshot();
  });

  it('should add a new task', () => {
    const { result } = renderHook(() => useTodos());
    act(() => {
      jest.runAllTimers();
    });

    act(() => {
      result.current.onAddTask('New Task');
    });

    expect(result.current.todos.length).toBe(1);
    expect(result.current.todos[0].title).toBe('New Task');
    expect(result.current.todos[0].isCompleted).toBe(false);
     expect(result.current).toMatchSnapshot();
  });

  it('should not add an empty task', () => {
    const { result } = renderHook(() => useTodos());
    act(() => {
      jest.runAllTimers();
    });

    act(() => {
      result.current.onAddTask('   ');
    });

    expect(result.current.todos.length).toBe(0);
    expect(result.current).toMatchSnapshot();
  });

  it('should delete a task', () => {
    const { result } = renderHook(() => useTodos());
    act(() => {
      jest.runAllTimers();
    });

    act(() => {
      result.current.onAddTask('Task to delete');
    });

    const id = result.current.todos[0].id;
    act(() => {
      result.current.onDeleteTask(id);
    });

    expect(result.current.todos.length).toBe(0);
    expect(result.current).toMatchSnapshot();
  });

  it('should toggle a task', () => {
    const { result } = renderHook(() => useTodos());
    act(() => {
      jest.runAllTimers();
    });

    act(() => {
      result.current.onAddTask('Toggle Task');
    });

    const id = result.current.todos[0].id;
    act(() => {
      result.current.onToggleTask(id);
    });

    expect(result.current.todos[0].isCompleted).toBe(true);

    act(() => {
      result.current.onToggleTask(id);
    });

    expect(result.current.todos[0].isCompleted).toBe(false);
    expect(result.current).toMatchSnapshot();
  });

  it('should save todos to storage when todos change', () => {
    const { result } = renderHook(() => useTodos());
    act(() => {
      jest.runAllTimers();
    });

    act(() => {
      result.current.onAddTask('Persist Task');
    });

    expect(Storage.saveToStorage).toHaveBeenCalledWith(STORAGE_KEY, expect.any(Array));
    expect(result.current).toMatchSnapshot();
  });
  it('should catch error when saving to storage fails', () => {
    (Storage.saveToStorage as jest.Mock).mockImplementationOnce(() => {
      throw new Error('Fake error');
    });

    const { result } = renderHook(() => useTodos());
    act(() => {
      jest.runAllTimers();
    });

    act(() => {
      result.current.onAddTask('Trigger Error');
    });
    expect(result.current).toMatchSnapshot();
  });
});
