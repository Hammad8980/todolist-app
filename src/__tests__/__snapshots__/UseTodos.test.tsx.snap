// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useTodos should add a new task 1`] = `
{
  "isLoading": false,
  "onAddTask": [Function],
  "onDeleteTask": [Function],
  "onToggleTask": [Function],
  "todos": [
    {
      "id": 1749654255943,
      "isCompleted": false,
      "priority": "p1",
      "title": "New Task",
    },
  ],
}
`;

exports[`useTodos should catch error when saving to storage fails 1`] = `
{
  "isLoading": false,
  "onAddTask": [Function],
  "onDeleteTask": [Function],
  "onToggleTask": [Function],
  "todos": [
    {
      "id": 1749654256027,
      "isCompleted": false,
      "priority": "p1",
      "title": "Trigger Error",
    },
  ],
}
`;

exports[`useTodos should delete a task 1`] = `
{
  "isLoading": false,
  "onAddTask": [Function],
  "onDeleteTask": [Function],
  "onToggleTask": [Function],
  "todos": [],
}
`;

exports[`useTodos should initialize with loading state 1`] = `
{
  "isLoading": true,
  "onAddTask": [Function],
  "onDeleteTask": [Function],
  "onToggleTask": [Function],
  "todos": [],
}
`;

exports[`useTodos should load todos from storage after timeout 1`] = `
{
  "isLoading": false,
  "onAddTask": [Function],
  "onDeleteTask": [Function],
  "onToggleTask": [Function],
  "todos": [
    {
      "id": 1,
      "isCompleted": false,
      "title": "Test",
    },
  ],
}
`;

exports[`useTodos should not add an empty task 1`] = `
{
  "isLoading": false,
  "onAddTask": [Function],
  "onDeleteTask": [Function],
  "onToggleTask": [Function],
  "todos": [],
}
`;

exports[`useTodos should save todos to storage when todos change 1`] = `
{
  "isLoading": false,
  "onAddTask": [Function],
  "onDeleteTask": [Function],
  "onToggleTask": [Function],
  "todos": [
    {
      "id": 1749654256017,
      "isCompleted": false,
      "priority": "p1",
      "title": "Persist Task",
    },
  ],
}
`;

exports[`useTodos should toggle a task 1`] = `
{
  "isLoading": false,
  "onAddTask": [Function],
  "onDeleteTask": [Function],
  "onToggleTask": [Function],
  "todos": [
    {
      "id": 1749654255995,
      "isCompleted": false,
      "priority": "p1",
      "title": "Toggle Task",
    },
  ],
}
`;
